lib/common/strings.c

```diff
@@ -363,52 +417,74 @@ pcmk__guint_from_hash(GHashTable *table, const char *key, guint default_val,
 long long
 crm_get_msec(const char *input)
 {
-    const char *num_start = NULL;
-    const char *units;
+    char *buf = NULL;
+    char *units = NULL; // Do not free; will point to part of buf
     long long multiplier = 1000;
     long long divisor = 1;
     long long msec = PCMK__PARSE_INT_DEFAULT;
-    size_t num_len = 0;
-    char *end_text = NULL;
 
     if (input == NULL) {
         return PCMK__PARSE_INT_DEFAULT;
     }
 
-    num_start = input + strspn(input, WHITESPACE);
-    num_len = strspn(num_start, NUMCHARS);
-    if (num_len < 1) {
+    buf = calloc(strlen(input) + 1, sizeof(char));
+    if (buf == NULL) {
         return PCMK__PARSE_INT_DEFAULT;
     }
-    units = num_start + num_len;
-    units += strspn(units, WHITESPACE);
 
-    if (!strncasecmp(units, "ms", 2) || !strncasecmp(units, "msec", 4)) {
+    clean_timeout_string(input, buf);
+    scan_ll(buf, &msec, -1, &units);
+
+    // Accept only nonnegative inputs
+    if (msec < 0) {
+        msec = PCMK__PARSE_INT_DEFAULT;
+        goto done;
+    }
+
+    /* @COMPAT Use exact comparisons. Currently, we match too liberally, and the
+     * second strncasecmp() in each case is redundant.
```

Comment:
I kept the second strncasecmp() in each case, so that we remember to add them to a `pcmk__strcase_any_of()` list later